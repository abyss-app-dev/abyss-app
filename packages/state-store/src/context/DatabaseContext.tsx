import type { SQliteClient } from '@abyss/records';
import { createContext, type ReactNode, useContext } from 'react';

interface DatabaseContextType {
    database: SQliteClient | null;
}

const DatabaseContext = createContext<DatabaseContextType>({ database: null });

interface DatabaseProviderProps {
    database: SQliteClient;
    children: ReactNode;
}

export function DatabaseProvider({ database, children }: DatabaseProviderProps) {
    return <DatabaseContext.Provider value={{ database }}>{children}</DatabaseContext.Provider>;
}

export function useDatabase(): SQliteClient {
    const context = useContext(DatabaseContext);
    if (!context.database) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context.database;
}
