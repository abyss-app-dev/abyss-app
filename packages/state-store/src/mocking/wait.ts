export function wait(ms = 50) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
