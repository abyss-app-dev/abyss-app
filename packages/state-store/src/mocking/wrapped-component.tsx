import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
