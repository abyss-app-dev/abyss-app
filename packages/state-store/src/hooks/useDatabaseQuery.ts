import type { SQliteClient } from '@abyss/records';
import { useCallback, useEffect, useRef } from 'react';
import { useDatabase } from '../context/DatabaseContext';
import { useQuery } from './useQuery';

export type DatabaseQueryFunction<R> = (database: SQliteClient) => R | Promise<R>;

export function useDatabaseQuery<R>(queryFunction: DatabaseQueryFunction<R>) {
    const database = useDatabase();

    // Use ref to store the latest query function to avoid dependency issues
    const queryFunctionRef = useRef(queryFunction);
    queryFunctionRef.current = queryFunction;

    // Stabilize the query function to prevent infinite re-renders
    const stableQueryFunction = useCallback((db: SQliteClient) => queryFunctionRef.current(db), []);

    const query = useQuery(async () => {
        return stableQueryFunction(database);
    }, [stableQueryFunction]);

    // Stabilize the subscription callback
    const handleSubscriptionUpdate = useCallback(async () => {
        const result = await stableQueryFunction(database);
        query.setData(result);
    }, [stableQueryFunction, database, query.setData]);

    useEffect(() => {
        const unsubscribe = database.subscribeDatabase(handleSubscriptionUpdate);
        return unsubscribe;
    }, [database, handleSubscriptionUpdate]);

    return query;
}
