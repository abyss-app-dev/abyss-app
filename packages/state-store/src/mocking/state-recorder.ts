import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
