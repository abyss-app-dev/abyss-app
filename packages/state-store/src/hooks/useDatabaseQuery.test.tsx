import { buildTestDB, type SQliteClient, SqliteTable } from '@abyss/records';
import { useEffect, useState } from 'react';
import { beforeEach, describe, test } from 'vitest';
import { DatabaseProvider } from '../context/DatabaseContext';
import { StateRecorder } from '../mocking/state-recorder';
import { TestRecorder } from '../mocking/wrapped-component';
import { useDatabaseQuery } from './useDatabaseQuery';

describe('useDatabaseQuery hook', () => {
    let db: SQliteClient;
    let recorder: StateRecorder;

    beforeEach(async () => {
        db = await buildTestDB();
        recorder = new StateRecorder();
    });

    test('applies query function to database', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseQuery(database => `Database: ${Object.keys(database.tables).length} tables`);
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('async query function on database', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseQuery(async database => {
                            await new Promise(resolve => setTimeout(resolve, 10));
                            return `Async DB: ${Object.keys(database.tables).length} tables`;
                        });
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('complex database query transformation', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseQuery(database => {
                            const tableNames = Object.keys(database.tables);
                            return {
                                tableCount: tableNames.length,
                                tableNames,
                                hasSettings: 'settings' in database.tables,
                                databaseType: 'sqlite',
                            };
                        });
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('query function with database access', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseQuery(async database => {
                            try {
                                const settingsRecord = await database.tables[SqliteTable.settings].default();
                                return `Found settings: ${settingsRecord.id}`;
                            } catch (error) {
                                return 'No settings found';
                            }
                        });
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('query function changes dynamically', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [version, setVersion] = useState(1);

                        const queryFn =
                            version === 1
                                ? (database: any) => `v1-tables: ${Object.keys(database.tables).length}`
                                : (database: any) => `v2-db-info: ${typeof database}`;

                        const result = useDatabaseQuery(queryFn);

                        useEffect(() => {
                            if (version === 1) {
                                setTimeout(() => setVersion(2));
                            }
                        }, [version]);

                        return { ...result, version };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manual refetch functionality', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasRefetched, setHasRefetched] = useState(false);

                        const result = useDatabaseQuery(database => `Refetch test: ${Object.keys(database.tables).length} tables`);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasRefetched) {
                                setHasRefetched(true);
                                setTimeout(() => result.refetch());
                            }
                        }, [result.data, result.loading, result.refetch, hasRefetched]);

                        return { ...result, hasRefetched };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription updates with database processing', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasSetData, setHasSetData] = useState(false);

                        const result = useDatabaseQuery(database => `Processed DB: ${Object.keys(database.tables).length} tables`);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasSetData && result.setData) {
                                setHasSetData(true);
                                setTimeout(() => result.setData('manually-set-database-data'));
                            }
                        }, [result.data, result.loading, result.setData, hasSetData]);

                        return { ...result, hasSetData };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription updates when database changes occur', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasTriggeredChange, setHasTriggeredChange] = useState(false);
                        const [queryResults, setQueryResults] = useState<string[]>([]);

                        const result = useDatabaseQuery(database => {
                            const tableCount = Object.keys(database.tables).length;
                            return `DB-Query: ${tableCount} tables at ${Date.now()}`;
                        });

                        // Track query results to see subscription updates
                        useEffect(() => {
                            if (result.data && !result.loading) {
                                setQueryResults(prev => [...prev, result.data as string]);
                            }
                        }, [result.data, result.loading]);

                        // Trigger a database change
                        useEffect(() => {
                            if (result.data && !result.loading && !hasTriggeredChange) {
                                setHasTriggeredChange(true);
                                setTimeout(async () => {
                                    // Creating a new record should trigger database subscription
                                    await db.tables[SqliteTable.settings].default();
                                }, 100);
                            }
                        }, [result.data, result.loading, hasTriggeredChange]);

                        return {
                            ...result,
                            hasTriggeredChange,
                            queryResultsCount: queryResults.length,
                            latestResult: queryResults[queryResults.length - 1],
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('multiple database queries receive independent updates', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [operationStep, setOperationStep] = useState(0);

                        const query1 = useDatabaseQuery(db => `Query1: ${Object.keys(db.tables).join(',')}`);
                        const query2 = useDatabaseQuery(db => `Query2: DB-${typeof db}`);

                        // Track when both queries are ready and trigger database change
                        useEffect(() => {
                            if (query1.data && query2.data && !query1.loading && !query2.loading && operationStep === 0) {
                                setOperationStep(1);
                                setTimeout(async () => {
                                    // Both queries should receive this update
                                    await db.tables[SqliteTable.settings].default();
                                    setOperationStep(2);
                                }, 100);
                            }
                        }, [query1.data, query1.loading, query2.data, query2.loading, operationStep]);

                        return {
                            query1Loading: query1.loading,
                            query2Loading: query2.loading,
                            query1HasData: !!query1.data,
                            query2HasData: !!query2.data,
                            operationStep,
                            bothReady: !!query1.data && !!query2.data && !query1.loading && !query2.loading,
                            query1Result: query1.data,
                            query2Result: query2.data,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('database query processes actual database state changes', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [operationPhase, setOperationPhase] = useState('initial');
                        const [recordCount, setRecordCount] = useState(0);

                        const result = useDatabaseQuery(async database => {
                            try {
                                // Get actual record from database
                                const defaultRecord = await database.tables[SqliteTable.settings].default();
                                return {
                                    hasDefaultRecord: !!defaultRecord,
                                    recordId: defaultRecord?.id,
                                    recordTheme: defaultRecord?.theme,
                                    queryTime: Date.now(),
                                };
                            } catch (error) {
                                return { error: 'Failed to query database', queryTime: Date.now() };
                            }
                        });

                        // Track how many times query runs
                        useEffect(() => {
                            if (result.data && !result.loading) {
                                setRecordCount(prev => prev + 1);
                            }
                        }, [result.data, result.loading]);

                        // Trigger database operations
                        useEffect(() => {
                            if (result.data && !result.loading && operationPhase === 'initial') {
                                setOperationPhase('triggering-change');
                                setTimeout(async () => {
                                    // Create another record to trigger subscription
                                    await db.tables[SqliteTable.settings].default();
                                    setOperationPhase('change-triggered');
                                }, 100);
                            }
                        }, [result.data, result.loading, operationPhase]);

                        return {
                            ...result,
                            operationPhase,
                            recordCount,
                            hasResult: !!result.data,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });
});
