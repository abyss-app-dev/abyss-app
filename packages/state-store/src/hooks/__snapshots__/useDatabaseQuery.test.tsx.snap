// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDatabaseQuery hook > applies query function to database 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Database: 11 tables",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseQuery hook > async query function on database 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Async DB: 11 tables",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseQuery hook > complex database query transformation 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "databaseType": "sqlite",
      "hasSettings": true,
      "tableCount": 11,
      "tableNames": [
        "settings",
        "modelConnection",
        "agentGraph",
        "messageThread",
        "metric",
        "message",
        "toolDefinition",
        "document",
        "chatSnapshot",
        "agentGraphExecution",
        "notebookCell",
      ],
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseQuery hook > database query processes actual database state changes 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasResult": false,
    "loading": true,
    "operationPhase": "initial",
    "recordCount": 0,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "hasDefaultRecord": true,
      "queryTime": 1748759093300,
      "recordId": "settings::default",
      "recordTheme": "",
    },
    "error": null,
    "hasResult": true,
    "loading": false,
    "operationPhase": "initial",
    "recordCount": 0,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "hasDefaultRecord": true,
      "queryTime": 1748759093300,
      "recordId": "settings::default",
      "recordTheme": "",
    },
    "error": null,
    "hasResult": true,
    "loading": false,
    "operationPhase": "triggering-change",
    "recordCount": 1,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseQuery hook > manual refetch functionality 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasRefetched": false,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Refetch test: 11 tables",
    "error": null,
    "hasRefetched": false,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Refetch test: 11 tables",
    "error": null,
    "hasRefetched": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Refetch test: 11 tables",
    "error": null,
    "hasRefetched": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseQuery hook > multiple database queries receive independent updates 1`] = `
[
  {
    "bothReady": false,
    "operationStep": 0,
    "query1HasData": false,
    "query1Loading": true,
    "query1Result": null,
    "query2HasData": false,
    "query2Loading": true,
    "query2Result": null,
  },
  {
    "bothReady": true,
    "operationStep": 0,
    "query1HasData": true,
    "query1Loading": false,
    "query1Result": "Query1: settings,modelConnection,agentGraph,messageThread,metric,message,toolDefinition,document,chatSnapshot,agentGraphExecution,notebookCell",
    "query2HasData": true,
    "query2Loading": false,
    "query2Result": "Query2: DB-object",
  },
  {
    "bothReady": true,
    "operationStep": 1,
    "query1HasData": true,
    "query1Loading": false,
    "query1Result": "Query1: settings,modelConnection,agentGraph,messageThread,metric,message,toolDefinition,document,chatSnapshot,agentGraphExecution,notebookCell",
    "query2HasData": true,
    "query2Loading": false,
    "query2Result": "Query2: DB-object",
  },
]
`;

exports[`useDatabaseQuery hook > query function changes dynamically 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
    "version": 1,
  },
  {
    "data": "v1-tables: 11",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "version": 1,
  },
  {
    "data": "v1-tables: 11",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "version": 2,
  },
]
`;

exports[`useDatabaseQuery hook > query function with database access 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Found settings: settings::default",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseQuery hook > subscription updates when database changes occur 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasTriggeredChange": false,
    "latestResult": undefined,
    "loading": true,
    "queryResultsCount": 0,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "DB-Query: 11 tables at 1748759092778",
    "error": null,
    "hasTriggeredChange": false,
    "latestResult": undefined,
    "loading": false,
    "queryResultsCount": 0,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "DB-Query: 11 tables at 1748759092778",
    "error": null,
    "hasTriggeredChange": true,
    "latestResult": "DB-Query: 11 tables at 1748759092778",
    "loading": false,
    "queryResultsCount": 1,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseQuery hook > subscription updates with database processing 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasSetData": false,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Processed DB: 11 tables",
    "error": null,
    "hasSetData": false,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Processed DB: 11 tables",
    "error": null,
    "hasSetData": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "manually-set-database-data",
    "error": null,
    "hasSetData": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;
